### DISCUSSION TITLE
How to run unit tests?

### DISCUSSION THREAD

**Discussion by octaeder** [texstudio-org/texstudio#2414] [Q&A] [answered]
I want to test utilsVersion.cpp. The test is utilsVersion_t.h. I believe I have to cutomize testmanager.cpp (comment lines with unwanted tests). But after build of txs with msys2 I can't find testmanager.o. What do I have to do?
Thanks!

**Comment by sunderme** [comment, marked as answer]
when you build a debug build (qmake CONFIG+=debug ...), the testmanager is build and automatically started on first run.
`--execute-tests`,`--execute-all-tests` and `--disable-tests` influence the behavior.

**Reply by octaeder** [reply]
Thanks, but something's wrong:

qmake6 CONFIG+=debug --execute-tests ../texstudio/texstudio.pro

Project MESSAGE: Building with Qt 6.3.0
Project MESSAGE: Static hunspell
Project MESSAGE: Static quazip5
Project MESSAGE: Creating debug version
Project MESSAGE: tests are enabled
Project MESSAGE: GIT bat

Then running make -j 8:

C:\msys64\home\WDAGUtilityAccount\texstudio-org\txsbuild_UpdateCheck/../texstudio/src/tests/syntaxcheck_t.cpp:301: undefined reference to `SyntaxCheck::waitForQueueProcess()'
error: ld returned 1 exit status
make: *** [Makefile:1194: texstudio.exe] Error 1

301: doc->SynChecker.waitForQueueProcess();

but syntaxcheck.cpp contains:

303: void SyntaxCheck::waitForQueueProcess(void)

**Comment by sunderme** [comment]
the `--execute-tests` etc. are command line options for starting texstudio, e.g. on the command line: `texstudio --execute-tests`.
So, the qmake command should be `qmake6 CONFIG+=debug ../texstudio/texstudio.pro`
In your case, I suspect, you need to do a `make clean` before beuild gain with `make -j 8`

**Reply by octaeder** [reply]
Thx for clarifying! Now I see that the results is written to a new document in open txs.

**Reply by octaeder** [reply]
can it be that the test hangs on the first test (smallusefullutils or so), when I run txs with test option after a normal run that ended with document tabs still open? After closing the tabs in a normal run and restart with test option the test started immediately. But I have to say that my very first test was the only one that succeded even so previous tabs where open. So for me it's not clear, where endless running comes from.

**Reply by sunderme** [reply]
I don't see hangs and I have always a document open.

**Reply by octaeder** [reply]
thx, I will retry.

**Reply by octaeder** [reply]
seems to be virus scanner. it checks new files at startup, takes about 15s. Looks a little bit frightening cause txs starts and after the scan it seems as if txs runs again. After stopping scanner it worked for me.

**Comment by octaeder** [comment]
I made a debug version from 4.3.0beta1-32-gc966dbca8 (current upstream/master) and run txs with --execute-tests. Result showed 17 Fails. Is this normal? A test of my version gives same result (none in updateCheck and utilsVersion). Just in case I give a screenshot:

![grafik](https://user-images.githubusercontent.com/102688820/174169946-c0c3440f-8793-4ceb-ab9d-99d4d087e2a2.png)

**Reply by sunderme** [reply]
the bidiMoving and latexCompleter you can ignore.
The rest should work, at least it does here and on CI.

**Reply by octaeder** [reply]
Random, this is a test of an already merged version, so this should have nothing to do with my changes. don't understand where this comes from.

**Reply by octaeder** [reply]
My latest commit of utilsversion_t.h (s. PR #2410) reports no Fails. So the question is why do some tests fail when I run the test?

First lines of output in msys2 console are

QList("C:\\Program Files\\", "C:\\Program Files (x86)\\", "C:/Program Files/", "C:/Program Files (x86)/")
inconsistent command arguments: "\\begin{crampedsubarray}" "mathtools.cwl"
inconsistent command arguments: "\\begin{subarray}" "amsmath.cwl"
inconsistent command arguments: "\\x" "kvdefinekeys.cwl"
inconsistent command arguments: "\\x" "ltxcmds.cwl"
inconsistent command arguments: "\\x" "infwarerr.cwl"
inconsistent command arguments: "\\x" "atbegshi.cwl"
inconsistent command arguments: "\\pgfmathrandom" "pgfmath.cwl"
inconsistent command arguments: "\\KOMAoptionsOf" "scrkbase.cwl"
inconsistent command arguments: "\\FamilyLengthMacroKey" "scrbase.cwl"
inconsistent command arguments: "\\FamilyLengthMacroKey" "scrbase.cwl"
inconsistent command arguments: "\\FamilyLengthMacroKey" "scrbase.cwl"
SmallUsefulFunctionsTest

I wonder, where these inconsistent commands errors come from. these do not appear in master. Maybe I should try with a new branch and PR.

**Reply by sunderme** [reply]
debug builds may give more internal information than release builds.
The inconsistent command arguments is a completer issue, typically harmless and can be ignored.
